integration/operator_test.go

```diff
@@ -0,0 +1,232 @@
+// +build k8srequired
+
+package integration
+
+import (
+	"fmt"
+	"io/ioutil"
+	"log"
+	"os"
+	"os/exec"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/cenk/backoff"
+	"github.com/giantswarm/certificatetpr"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	v1 "k8s.io/client-go/pkg/api/v1"
+)
+
+const (
+	certOperatorValuesFile = "/tmp/cert-operator-install.yaml"
+	defaultDeadline        = 15
+)
+
+var cs kubernetes.Interface
+
+// TestMain allows us to have common setup and teardown steps that are run
+// once for all the tests https://golang.org/pkg/testing/#hdr-Main.
+func TestMain(m *testing.M) {
+	var v int
+	var err error
+	cs, err = getK8sClient()
+	if err != nil {
+		v = 1
+		log.Printf("unexpected error: %v\n", err)
+	}
+
+	if err := setUp(cs); err != nil {
+		v = 1
+		log.Printf("unexpected error: %v\n", err)
+	}
+
+	if v == 0 {
+		v = m.Run()
+	}
+
+	tearDown(cs)
+
+	os.Exit(v)
+}
+
+func TestSecretsAreCreated(t *testing.T) {
+	err := runCmd("helm registry install quay.io/giantswarm/cert-resource-lab-chart:stable -- -n cert-resource-lab --set commonDomain=${COMMON_DOMAIN} --set clusterName=${CLUSTER_NAME}")
+	if err != nil {
+		t.Errorf("could not install cert-resource-lab, %v", err)
+	}
+
+	secretName := fmt.Sprintf("%s-api", os.Getenv("CLUSTER_NAME"))
+	err = waitFor(secretFunc(cs, "default", secretName))
+	if err != nil {
+		t.Errorf("could not find expected secret: %v", err)
+	}
+}
+
+func setUp(cs kubernetes.Interface) error {
+	if err := createGSNamespace(cs); err != nil {
+		return err
+	}
+
+	if err := installVault(cs); err != nil {
+		return err
+	}
+
+	if err := installCertOperator(cs); err != nil {
+		return err
+	}
+	return nil
+}
+
+func tearDown(cs kubernetes.Interface) {
+	runCmd("helm delete vault --purge")
+	runCmd("helm delete cert-operator --purge")
+	runCmd("helm delete cert-resource-lab --purge")
+	cs.CoreV1().Namespaces().Delete("giantswarm", &metav1.DeleteOptions{})
+	cs.ExtensionsV1beta1().ThirdPartyResources().Delete(certificatetpr.Name, &metav1.DeleteOptions{})
+}
+
+func createGSNamespace(cs kubernetes.Interface) error {
+	// check if the namespace already exists
+	_, err := cs.CoreV1().Namespaces().Get("giantswarm", metav1.GetOptions{})
+	if err == nil {
+		return nil
+	}
+
+	namespace := &v1.Namespace{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "giantswarm",
+		},
+	}
+	_, err = cs.CoreV1().Namespaces().Create(namespace)
+	if err != nil {
+		return err
+	}
+
+	return waitFor(activeNamespaceFunc(cs, "giantswarm"))
+}
+
+func installVault(cs kubernetes.Interface) error {
+	if err := runCmd("helm registry install quay.io/giantswarm/vaultlab-chart:stable -- --set vaultToken=${VAULT_TOKEN} -n vault"); err != nil {
+		return err
+	}
+
+	return waitFor(runningPodFunc(cs, "default", "app=vault"))
+}
+
+func installCertOperator(cs kubernetes.Interface) error {
+	installationValues := `commonDomain: ${COMMON_DOMAIN}
```

Comment:
It would be nice to have a constant for this template and a godoc.

Reply:
Sure, done.